pkg/controllers/sync/controller.go

```diff
@@ -1104,58 +1111,120 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
-		// Otherwise, we will have to issue a list request.
-		if hasSynced() {
-			objects, err := resourceLister.List(labels.Everything())
+			// get all managed cluster objects according to ftc type
+			clusterObjs, namespaces, err := s.listManagedClusterObjects(ctx, ftc, cluster.Name)
 			if err != nil {
-				remainingByGVK[gvk] = fmt.Sprintf("Unknown (failed to list from cluster lister: %v)", err)
-			} else if len(objects) > 0 {
-				remainingByGVK[gvk] = strconv.Itoa(len(objects))
+				logger.Error(err, "Failed to list clusterObjects")
+				syncMap.Store(gvk, fmt.Sprintf("failed to list cluster objects for %s", gvk))
+				return
 			}
-		} else {
-			client, exists := s.fedInformerManager.GetClusterDynamicClient(cluster.Name)
-			if !exists {
-				remainingByGVK[gvk] = "Unknown (cluster client does not exist)"
-				continue
+			if len(clusterObjs) == 0 {
+				return
 			}
 
-			objects, err := client.Resource(ftc.GetSourceTypeGVR()).Namespace(corev1.NamespaceAll).List(
-				ctx, metav1.ListOptions{
-					Limit: 1,
-					LabelSelector: labels.SelectorFromSet(labels.Set{
-						managedlabel.ManagedByKubeAdmiralLabelKey: managedlabel.ManagedByKubeAdmiralLabelValue,
-					}).String(),
-				},
-			)
-			if err == nil && len(objects.Items) > 0 {
-				remainingByGVK[gvk] = strconv.Itoa(len(objects.Items))
-			} else if err != nil && !meta.IsNoMatchError(err) && !apierrors.IsNotFound(err) {
-				remainingByGVK[gvk] = fmt.Sprintf("Unknown (failed to list from cluster: %v)", err)
+			client, err := s.getClusterClient(cluster.Name)
+			if err != nil {
+				logger.Error(err, "Failed to get cluster client")
+				syncMap.Store(gvk, "failed to get cluster client")
+				return
 			}
-		}
+
+			// update some cluster objects before deleting them
+			var updateFailedObjs []string
+			for i := range clusterObjs {
+				clusterObj := clusterObjs[i].DeepCopy()
+				keyedLogger := logger.WithValues(
+					"namespace", clusterObj.GetNamespace(),
+					"name", clusterObj.GetName(),
+					"gvk", gvk)
+				namespacedName := fmt.Sprintf("%s/%s", clusterObj.GetNamespace(), clusterObj.GetName())
+
+				// get the corresponding federated object
+				fedObjName := naming.GenerateFederatedObjectName(clusterObj.GetName(), ftc.Name)
+				fedObj, err := s.fedAccessor.FederatedResource(common.QualifiedName{
+					Namespace: clusterObj.GetNamespace(),
+					Name:      fedObjName,
+				})
+				if err != nil {
+					keyedLogger.Error(err, "Failed to get corresponding federated object")
+					updateFailedObjs = append(updateFailedObjs, namespacedName)
+					continue
+				}
+				if fedObj == nil {
+					err := fmt.Errorf("the federated object %s is not found", fedObjName)
+					keyedLogger.Error(err, "Failed to get corresponding federated object")
+					updateFailedObjs = append(updateFailedObjs, namespacedName)
+					continue
+				}
+
+				// remove the finalizer added by admiral
+				needUpdate, err := dispatch.RemoveRetainObjectFinalizer(clusterObj)
```

Comment:
Should this method be extracted out from the dispatch package?

Reply:
The method has been extracted.